### ISSUE TITLE
Running the model locally

### ISSUE THREAD

**Issue by jyotikhetan** [jina-ai/jina#4073]
**I want to run the model locally**
 how can run any model locally, without pulling anything from hub..!! I tried cross modal replaced jinahub model by python script.. still it's connecting to hub to load.. process the thing..!! 

My Flow look like tihs now 
````
jtype: Flow                                      
version: '1'                                    
with:                                             
  prefetch: 10                                    
  port_expose: $JINA_PORT                         
  workspace: $JINA_WORKSPACE                      
pods:                                             
  - name: image_loader                            
    uses: ImageReader                            
    py_modules: 'flows/executors.py'              
    read_only: true                             
    needs: gateway                                
  - name: image_encoder                           
    uses : CLIPImageEncoder
    py_modules: 'flows/clip_image_encoder.py'                           
    needs: image_loader                           
  - name: image_indexer                           
    uses : SimpleIndexer
    py_modules: 'flows/simple_indexer.py'
         
    uses_with:                                    
      index_file_name: 'image_index'             
    needs: image_encoder                          
  - name: text_filter                            
    uses: TextFilter                             
    py_modules: 'flows/executors.py'             
    needs: gateway                              
  - name: text_encoder                          
    uses : CLIPTextEncoder
    py_modules: 'flows/clip_text_encoder.py'                        
    needs: text_filter                           
  - name: text_indexer                          
    uses: SimpleIndexer
    py_modules: 'flows/simple_indexer.py'
    uses_with:                                 
      index_file_name: 'text_index'              
    needs: text_encoder                           
  - name: join_all                             
    needs: [image_indexer, text_indexer]
```


**Please help me to understand, How I can proceed further on this**

**Comment by JoanFM** [comment]
Hey @jyotikhetan , are you sure there is no other `Flow` also involved. There should be no `jinahub` involvement here.

Can you share what is the code u run and what are the logs u see?

Thanks,

Joan

**Comment by jyotikhetan** [comment]
hi @JoanFM , so I am running the executor file from jina... For CLIPtextencoder, I have removed path of executor from jinahub & replaced it with python script..as below 

```
```
`from typing import Dict, Optional, Sequence

import torch
from jina import DocumentArray
from transformers import CLIPModel, CLIPTokenizer


class CLIPTextEncoder():
 
    def __init__(
        self,
        pretrained_model_name_or_path: str = 'openai/clip-vit-base-patch32',
        base_tokenizer_model: Optional[str] = None,
        max_length: int = 77,
        device: str = 'cpu',
        traversal_paths: Sequence[str] = ['r'],
        batch_size: int = 32,
        *args,
        **kwargs,
    ):
    
        super().__init__(*args, **kwargs)
        self.traversal_paths = traversal_paths
        self.batch_size = batch_size
        self.pretrained_model_name_or_path = pretrained_model_name_or_path
        self.base_tokenizer_model = (
            base_tokenizer_model or pretrained_model_name_or_path
        )
        self.max_length = max_length

        self.device = device
        self.tokenizer = CLIPTokenizer.from_pretrained(self.base_tokenizer_model)
        self.model = CLIPModel.from_pretrained(self.pretrained_model_name_or_path)
        self.model.eval().to(device)

    
    def encode(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
     
        if docs is None:
            return

        for docs_batch in docs.batch(
            traversal_paths=parameters.get('traversal_paths', self.traversal_paths),
            batch_size=parameters.get('batch_size', self.batch_size),
            require_attr='text',
        ):
            text_batch = docs_batch.get_attributes('text')

            with torch.inference_mode():
                input_tokens = self._generate_input_tokens(text_batch)
                embeddings = self.model.get_text_features(**input_tokens).cpu().numpy()
                for doc, embedding in zip(docs_batch, embeddings):
                    doc.embedding = embedding

    def _generate_input_tokens(self, texts: Sequence[str]):

        input_tokens = self.tokenizer(
            texts,
            max_length=self.max_length,
            padding='longest',
            truncation=True,
            return_tensors='pt',
        )
        input_tokens = {k: v.to(self.device) for k, v in input_tokens.items()}
        return input_tokens`
```

I am running the same python app.py script file given in repo... I am getting this error while running it..

`                 (raised from /home/jyoti/making_py/makingpyenv/lib/python3.7/site-packages/jina/importer.py:197)
  image_encoder@7928[C]:can not load the executor from CLIPImageEncoder
  image_encoder@7928[E]:ExecutorFailToLoad() during <class 'jina.peapods.runtimes.zmq.zed.ZEDRuntime'> initialization
 add "--quiet-error" to suppress the exception details
`
```

**Comment by JoanFM** [comment]
hey @jyotikhetan ,

his is not `Jinahub` problem. The problem is your `Executor` is not inherting from `Executor` and is not using` @requests` decorator

**Comment by JoanFM** [comment]
```python
from typing import Dict, Optional, Sequence

import torch
from jina import DocumentArray, Executor, requests
from transformers import CLIPModel, CLIPTokenizer


class CLIPTextEncoder(Executor):
 
    def __init__(
        self,
        pretrained_model_name_or_path: str = 'openai/clip-vit-base-patch32',
        base_tokenizer_model: Optional[str] = None,
        max_length: int = 77,
        device: str = 'cpu',
        traversal_paths: Sequence[str] = ['r'],
        batch_size: int = 32,
        *args,
        **kwargs,
    ):
    
        super().__init__(*args, **kwargs)
        self.traversal_paths = traversal_paths
        self.batch_size = batch_size
        self.pretrained_model_name_or_path = pretrained_model_name_or_path
        self.base_tokenizer_model = (
            base_tokenizer_model or pretrained_model_name_or_path
        )
        self.max_length = max_length

        self.device = device
        self.tokenizer = CLIPTokenizer.from_pretrained(self.base_tokenizer_model)
        self.model = CLIPModel.from_pretrained(self.pretrained_model_name_or_path)
        self.model.eval().to(device)

    @requests
    def encode(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
     
        if docs is None:
            return

        for docs_batch in docs.batch(
            traversal_paths=parameters.get('traversal_paths', self.traversal_paths),
            batch_size=parameters.get('batch_size', self.batch_size),
            require_attr='text',
        ):
            text_batch = docs_batch.get_attributes('text')

            with torch.inference_mode():
                input_tokens = self._generate_input_tokens(text_batch)
                embeddings = self.model.get_text_features(**input_tokens).cpu().numpy()
                for doc, embedding in zip(docs_batch, embeddings):
                    doc.embedding = embedding

    def _generate_input_tokens(self, texts: Sequence[str]):

        input_tokens = self.tokenizer(
            texts,
            max_length=self.max_length,
            padding='longest',
            truncation=True,
            return_tensors='pt',
        )
        input_tokens = {k: v.to(self.device) for k, v in input_tokens.items()}
        return input_tokens`

```

**Comment by JoanFM** [comment]
I guess there is some other problem with `ClipImageEncoder`

**Comment by jyotikhetan** [comment]
hi @JoanFM  Thank you for response..!But I tried with simpleindexer, I still couldn't make it work..!! Basically I want to run my model completely locally, in sense I should be able to run it without internet connection....What i need to do for that..!!

**Comment by JoanFM** [comment]
You can try to do `jina hub pull` of the `Executors` you are interested to use. And then have the `local` copy of these `python` files to deploy as normal `Executors`. Another option is to clone locally the `executors` from their `public` repositories, and just use the `code` as local executors.

**Comment by jyotikhetan** [comment]
I tried the second option with cross modal... replaced all the executor with python script of corresponding executor... with internet it's working, without internet it's not...!! Excuse me for such naive query... but I am trying to understand..!!

**Comment by JoanFM** [comment]
I need to have more details `logs`. What is the `exact` thing you run and the exact problem?

**Comment by jyotikhetan** [comment]
I am attaching the flow folder & script I am trying to run... I want to run everything locally...!! In sense , I should be able to run my model without network connectivity.. nothing should  be downloaded from hub or cloud.... If i can do that...??
[app.txt](https://github.com/jina-ai/jina/files/7735226/app.txt)

[flows.zip](https://github.com/jina-ai/jina/files/7735214/flows.zip)

**Comment by JoanFM** [comment]
You have to change the `flow-query.yml` to use the `non-jinahub` executors.

**Comment by jyotikhetan** [comment]
So Can I just create my own Imagencoder executor or dockerize the hub executor... make the changes in flow accordingly...then will it work??

**Comment by JoanFM** [comment]
The changes that u did to your `flow-index.yml` should also be done in `flow-query.yml`

**Comment by jyotikhetan** [comment]
Okay... But I haven't used query flow... I was just trying to index first..!! So I am able to run the above flow-index.yml with network connectivity but without it it's not able to index....

**Comment by JoanFM** [comment]
Can u tell me what is the exact logs when u run the `index-flow`?

I think is not about `jinahub` , maybe is it truing to download the transformer models?

**Comment by jyotikhetan** [comment]
hey hi @JoanFM, Thank you for your support..!! It's working now....I had an issue in loading the transformer model as you correctly mentioned...!!
